### DISCUSSION TITLE
How to add a Diagnostic to a DiagnosticCollection: Error: Can't resolve '@theia/plugin'

### DISCUSSION THREAD

**Discussion by scharf** [eclipse-theia/theia#14389] [Q&A] [answered]
Here is my naive simple code. I want to add errors to a file (without LSP). Just from within my code.

```typescript
import {injectable} from 'inversify';
import {Diagnostic, DiagnosticCollection, DiagnosticSeverity, languages, Position, Range, Uri} from '@theia/plugin';

@injectable()
export class ProblemContributor {
    private diagnosticsCollection: DiagnosticCollection;

    constructor() {
        this.diagnosticsCollection = languages.createDiagnosticCollection('formula-editor');
    }

    addProblem(filePath: string, line: number, message: string) {
        const uri = Uri.file(filePath);
        const range = new Range(new Position(line, 0), new Position(line, 100)); // Adjust as needed
        const diagnostic = new Diagnostic(range, message, DiagnosticSeverity.Error);

        this.diagnosticsCollection.set(uri, [diagnostic]);
    }
}
```

I tried it in node and in the browser. Adding `@theia/plugin` 

I get:

```
ERROR in ../../theia-extensions/formula-runner/lib/browser/problems/ProblemContributor.js 14:17-41
Module not found: Error: Can't resolve '@theia/plugin' in '/MyPath/lib/browser/problems'
```

Then I added `"@theia/languages"`

And I get
```
ERROR in ../../node_modules/monaco-languageclient/lib/connection.js 7:15-65
Module not found: Error: Can't resolve 'vscode-languageserver-protocol/lib/main'
```

Then I added `"vscode-languageserver-protocol"`

Lots of errors. The I tried:

        "@theia/languages": "^1.4.0",
        "@theia/plugin": "^1.55.0",
        "monaco-languageclient": "^8.8.3",
        "vscode-ws-jsonrpc": "^3.3.2"
        
Now I get the following build Error

```

ERROR in ../../node_modules/@theia/languages/lib/node/language-server-contribution.js 84:15-54
Module not found: Error: Package path ./lib/server is not exported from package 
 @ ../../node_modules/@theia/languages/lib/node/languages-backend-module.js 22:37-78
 @ ./src-gen/backend/server.js 79:19-80
 @ ./src-gen/backend/main.js 248:21-40

```

...and many more variations -- each one leading to another error.

I cannot figure out what dependency I am missing.

**Comment by scharf** [comment]
may be related to https://github.com/TypeFox/monaco-languageclient/issues/242

**Comment by scharf** [comment]
...many hours later....

it boild down to

```
 Can't resolve '@theia/plugin'
 ```

I simply cannot figure out how to use anything in ` @theia/plugin`. I am sure there is somewhere a documentation on how to use it.

I guess I have to do some magic inject or something else...

Google is not your friend when it comes to searching solutions for theia....

**Comment by scharf** [comment]
It would be **very** helpful to have at least a hint small of how to use the API in the [packages/plugin/README.md](https://github.com/eclipse-theia/theia/blob/master/packages/plugin/README.md).

There are plenty of documents and videos describing all kinds of concepts, but a basic code snippet (including imports and the required packages) would be **extremely** helpful to explain how to consume anything in `@theia/plugin`.

In the [packages/plugin/README.md](https://github.com/eclipse-theia/theia/blob/master/packages/plugin/README.md) there is a basic snippet, but it does not explain how I get to the variable `theia` (what to import, what to install) and what this is.

```typescript
theia.commands.registerCommand({id:'say.hello.command'}, ()=>{
    console.log("Hello World!");
});
```

**Comment by msujew** [comment, marked as answer]
Hey @scharf,

The `@theia/plugin` package/API isn't designed to be used by Theia applications, but rather Theia plugins (more commonly known as VS Code extensions).

You could either write a VS Code extension that contributes that feature to your IDE (the `Diagnostic` API is available there without going through the hoops of building a language server) and include that in your Theia app, or use the underlying Theia API (which would be the [`@theia/markers`](https://github.com/eclipse-theia/theia/tree/master/packages/markers) package) in your Theia extension.
